toolchain/semantics/semantics_context.h

```diff
@@ -233,10 +233,17 @@ class SemanticsContext {
   auto CanonicalizeStructType(ParseTree::Node parse_node,
                               SemanticsNodeBlockId refs_id) -> SemanticsTypeId;
 
+  auto CanonicalizeTupleType(ParseTree::Node parse_node,
+                             const llvm::SmallVector<SemanticsTypeId>& type_ids)
+      -> SemanticsTypeId;
   // Converts an expression for use as a type.
   // TODO: This should eventually return a type ID.
   auto ExpressionAsType(ParseTree::Node parse_node, SemanticsNodeId value_id)
       -> SemanticsTypeId {
+    auto node = semantics_ir_->GetNode(value_id);
+    if (node.kind() == SemanticsNodeKind::StubReference) {
+      value_id = node.GetAsStubReference();
+    }
```

Comment:
Same comment as before: if we can form a `StubReference` to a `StubReference` we'll need a loop here.